File: my_code/DSA/JS/DS/linked-lists/singly-linked-list/reverse-linked-list [recursive method].js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
  }

  push(val) {
    const newNode = new Node(val);
    if (!this.head) {
      this.head = newNode;
      return this;
    }
    let current = this.head;
    while (current.next) {
      current = current.next;
    }
    current.next = newNode;
    return this;
  }

  display() {
    let current = this.head;
    let output = "";
    while (current) {
      output += current.val + " -> ";
      current = current.next;
    }
    output += "None";
    console.log(output);
  }

  reverseList(head) {
    // Base Case
    if (!head) {
      return null;
    }

    let current = head;
    // each time consider the current as a head
    if (current.next) {
      // call the recursive function
      current = this.reverseList(head.next);
      // set the head.next point to the head itself
      // node1 => node2
      // node2 => node1
      head.next.next = head;
    }
    // Break the original link
    head.next = null;

    return current;
  }

  reverse() {
    this.head = this.reverseList(this.head);
  }
}

const myList = new LinkedList();
[1, 2, 3, 4, 5].forEach((n) => myList.push(n));

console.log("Original List:");
myList.display();

myList.reverse();

console.log("Reversed List:");
myList.display();

// Recursive
// T => O(n), M => O(n)

File: my_code/DSA/PYTHON/DS/linked lists/singly-linked-list/reverse-linked-list [recursive method].py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def push(self, val):
        new_node = Node(val)
        if not self.head:
            self.head = new_node
            return self
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node
        return self

    def display(self):
        current = self.head
        while current:
            print(current.val, end=" -> ")
            current = current.next
        print("None")

    def reverseList(self, head):
        # Base Case
        if not head:
            return None
        
        current = head
        # each time consider the current as a head
        if current.next:
            # call the recursive function
            current = self.reverseList(head.next)
            # set the head.next point to the head itself
            # node1 => node2
            # node2 => node1
            head.next.next = head
        # Break the original link
        head.next = None

        return current

    def reverse(self):
        self.head = self.reverseList(self.head)



my_list = LinkedList()
for n in [1, 2, 3, 4, 5]:
    my_list.push(n)

print("Original List:")
my_list.display()

my_list.reverse()

print("Reversed List:")
my_list.display()


# Recurcive
# T O(n), M O(n)